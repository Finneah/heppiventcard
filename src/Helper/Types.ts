export type StampCardType = {
  id?: string;
  title: string;
  complete: Boolean;
  date_of_creation: Date;
  date_of_completed?: Date;
  completed_image?: string;
};

export interface SectionPart extends StampCardType {
  content: StampType[] | null | undefined;
}

export type StampType = {
  id?: string;
  number: number;
  name: string | null | undefined;
  image: string | null | undefined;
  picture: string | null | undefined;
  done: Boolean;
  date: Date | undefined;
  description: string | null | undefined;
  stampCard: StampCardType;
};

export type UserType = {
  id?: number;
  name: string | null | undefined;
  rank: number;
};

export type QRCodeObject = {
  url: string;
  date: string;
  description: string;
  name: string;
  picture: string | null | undefined;
};
export type QRCodeReturnData = {
  image: string;
  description: string;
  date: Date;
  done: boolean;
  name: string;
};
