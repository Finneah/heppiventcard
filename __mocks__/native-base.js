jest.mock('native-base', () => {
  return {
    StyleProvider: jest.fn().mockImplementation((x) => x),
  };
});
