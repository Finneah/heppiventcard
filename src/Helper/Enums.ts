export enum DEBUGLEVEL {
  OFF,
  DATA,
  ALL,
}
export enum UserRank {
  RANK_00 = 0,
  RANK_03 = 3,
  RANK_10 = 10,
  RANK_15 = 15,
  RANK_20 = 20,
  RANK_30 = 30,
}
