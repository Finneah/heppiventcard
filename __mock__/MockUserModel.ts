import {UserType} from '../src/Helper/Types';

/**
 * @category Database
 * @class UserModel
 * @classdesc DB Model for User
 */
export default class MockUserModel {
  /**
   * @memberof UserModel
   * @description get Userobject from DB
   * there is only one User
   * @param {Object} user
   */
  getUser = (user: object): any => {
    return user;
  };

  createNewUser = (user: UserType): void => {
    console.log(user, 'User.insert(user, true)');
  };

  updateUser = (user: UserType): void => {
    console.log(user, 'User.update(user, user, true)');
  };
}
