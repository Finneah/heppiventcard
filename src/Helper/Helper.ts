import {StampCards, Stamps} from '../database';
import {UserRank} from './Enums';
import {strings} from '../locale/i18n';
import {SectionPart, StampCardType, UserType} from './Types';
import moment from 'moment';
import 'moment/locale/de';
export const isValidUser = (user: UserType): Boolean => {
  try {
    if (user?.name !== '' || user?.rank >= 0) {
      return true;
    }
    throw new Error('name or rank is missing');
  } catch (error) {
    return false;
  }
};
/**
 * "RANK_0": "Gast",
 * "RANK_3": "Stammgast",
 * "RANK_10": "Mitglied",
 * "RANK_15": "Stammmitglied",
 * "RANK_20": "Platinmitglied",
 * "RANK_30": "Goldmitglied",
 */
export const _getRank = (user: UserType): string => {
  let returnValue = strings('RANK_00');
  if (user?.rank) {
    if (user.rank >= UserRank.RANK_30) {
      returnValue = strings('RANK_30');
    } else if (user.rank >= UserRank.RANK_20) {
      returnValue = strings('RANK_20');
    } else if (user.rank >= UserRank.RANK_15) {
      returnValue = strings('RANK_15');
    } else if (user.rank >= UserRank.RANK_10) {
      returnValue = strings('RANK_10');
    } else if (user.rank >= UserRank.RANK_03) {
      returnValue = strings('RANK_03');
    } else {
      returnValue = strings('RANK_00');
    }
  }
  return returnValue;
};

export const _getImageNameForRank = (user: UserType): string => {
  if (user?.rank >= 30) {
    return 'Gold';
  } else if (user.rank >= 20) {
    return 'Platin';
  }
  return 'Standard';
};

export const _checkDoneItems = (stampCard: SectionPart): number | string => {
  try {
    var done = 0;

    if (stampCard?.content) {
      stampCard?.content?.forEach((content: any) => {
        if (content.done === 1 || content.done === true) {
          done++;
        }
      });

      return done;
    } else {
      throw new Error('Error on _checkDoneItems');
    }
  } catch (error) {
    console.warn('_checkDoneItems', error);
    return '_checkDoneItems' + ' ' + error.message;
  }
};

export const _getStampCardTitle = (stampCard: SectionPart): string => {
  let firstDate: Date | undefined;
  let lastDate: Date | undefined;
  let title = '';

  if (stampCard?.content) {
    for (const element of stampCard.content) {
      if (!firstDate && element.date) {
        firstDate = element.date;
      } else {
        if (
          firstDate &&
          element?.date !== null &&
          element?.date !== undefined &&
          element?.date < firstDate
        ) {
          firstDate = element.date;
        }
      }
      if (!lastDate) {
        lastDate = element.date;
      } else {
        if (
          lastDate &&
          element?.date !== null &&
          element?.date !== undefined &&
          element?.date > lastDate
        ) {
          lastDate = element.date;
        }
      }

      if (firstDate && lastDate) {
        title =
          moment(firstDate).locale('de').format('L') +
          ' - ' +
          moment(lastDate).locale('de').format('L');
        console.log(title);
      }
    }
  } else {
    return title;
  }

  return title;
};

export const _createNewStampCard = async () => {
  var newStampCard: StampCardType = {
    title: '',
    complete: false,
    date_of_creation: new Date(),
    date_of_completed: undefined,
    completed_image: undefined,
  };

  newStampCard.date_of_creation = new Date();
  newStampCard.title = '';
  newStampCard.complete = false;

  var card = await StampCards.insert(newStampCard, true)[0];

  var newStamps = [];
  for (let i = 1; i <= 10; i++) {
    newStamps.push({
      number: i,
      done: 0,
      stampCard: card,
    });
  }

  await Stamps.insert(newStamps, true);
};
