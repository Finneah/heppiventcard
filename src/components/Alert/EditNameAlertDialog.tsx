import {
  AlertDialog,
  Button,
  FormControl,
  Icon,
  Input,
  KeyboardAvoidingView,
  Stack,
} from 'native-base';

import * as React from 'react';
import {UserModel} from '../../database/Models/UserModel';
import {UserType} from '../../Helper/Types';
import {strings} from '../../locale/i18n';
import Ionicons from 'react-native-vector-icons/Ionicons';
let userModel = new UserModel();
type Props = {
  user: UserType;
  showAlertDialog: boolean;
  setShowAlertDialog: Function;
};

const EditNameAlertDialog: React.FC<Props> = ({
  user,
  showAlertDialog = false,
  setShowAlertDialog,
}) => {
  const onClose = () => setShowAlertDialog(false);
  const inputRef = React.useRef();
  const [name, setName] = React.useState('');
  const [isInvalid, setIsInvalid] = React.useState(false);
  React.useEffect(() => {
    setIsInvalid(name.length < 3 || name.length > 8);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  React.useEffect(() => {
    setIsInvalid(name.length < 3 || name.length > 8);
  }, [name]);
  return showAlertDialog ? (
    <KeyboardAvoidingView>
      <AlertDialog
        leastDestructiveRef={inputRef}
        onClose={onClose}
        isOpen={showAlertDialog}>
        <AlertDialog.Content>
          <AlertDialog.CloseButton
            icon={<Icon color="primary.600" as={<Ionicons name="close" />} />}
          />
          <AlertDialog.Header>
            {strings('CHANGE_NAME_TITLE')}
          </AlertDialog.Header>
          <AlertDialog.Body>
            <FormControl isRequired isInvalid={isInvalid}>
              <Stack mx={4}>
                <Input
                  p={2}
                  placeholder="Anzeigename"
                  ref={inputRef}
                  onChangeText={(text) => {
                    setName(text);
                  }}
                />

                <FormControl.ErrorMessage>
                  {'Mindestens 3 Zeichnen und maximal 8 Zeichnen'}
                </FormControl.ErrorMessage>
              </Stack>
            </FormControl>
          </AlertDialog.Body>
          <AlertDialog.Footer>
            <Button
              colorScheme="primary"
              ml={3}
              disabled={isInvalid}
              onPress={() => {
                user.name = name;
                userModel.updateUser(user);
                onClose();
              }}>
              {strings('SAVE')}
            </Button>
          </AlertDialog.Footer>
        </AlertDialog.Content>
      </AlertDialog>
    </KeyboardAvoidingView>
  ) : null;
};

export default EditNameAlertDialog;
