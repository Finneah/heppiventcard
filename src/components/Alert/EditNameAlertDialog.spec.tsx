import 'react-native';
import * as React from 'react';
import {fireEvent, render} from '@testing-library/react-native';

// Note: test renderer must be required after react-native.

import {NativeBaseProvider} from 'native-base';

import EditNameAlertDialog from './EditNameAlertDialog';
import {UserType} from '../../Helper/Types';
import MockUserModel from './../../../__mock__/MockUserModel';
// let mockUserModel = new MockUserModel();
let baseUser: UserType = {
  name: 'Jenni',
  rank: 0,
};

jest.mock('./../../../__mock__/MockUserModel.ts', () => MockUserModel);
const wrapper = ({children}: any) => (
  <NativeBaseProvider
    initialWindowMetrics={{
      frame: {x: 0, y: 0, width: 0, height: 0},
      insets: {top: 0, left: 0, right: 0, bottom: 0},
    }}>
    {children}
  </NativeBaseProvider>
);

describe('MainAlert', () => {
  test('renders correctly', () => {
    const editNameAlertDialog = render(
      <EditNameAlertDialog
        user={baseUser}
        showAlertDialog={true}
        setShowAlertDialog={jest.fn()}
      />,
      {
        wrapper,
      },
    );

    expect(editNameAlertDialog.toJSON()).toMatchSnapshot();
  });
  test('FireEvent correctly', () => {
    const {getByText, getByPlaceholderText} = render(
      <EditNameAlertDialog
        user={baseUser}
        showAlertDialog={true}
        setShowAlertDialog={jest.fn()}
      />,
      {wrapper},
    );

    const saveButton = getByText('SAVE');
    expect(saveButton).toBeTruthy();
    if (saveButton) {
      fireEvent.press(saveButton);
    }
    const input = getByPlaceholderText('Anzeigename');
    expect(input).toBeTruthy();
    if (input) {
      fireEvent.changeText(input, 'Jenni');
    }
  });
});
