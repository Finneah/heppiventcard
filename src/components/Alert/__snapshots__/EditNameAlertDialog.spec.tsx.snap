// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainAlert renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {},
      ]
    }
  />
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      needsOffscreenAlphaCompositing={true}
      pointerEvents="box-none"
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "transform": Array [
            Object {
              "translateY": 0,
            },
            Object {
              "translateX": 0,
            },
            Object {
              "scale": 1,
            },
            Object {
              "scaleX": 1,
            },
            Object {
              "scaleY": 1,
            },
            Object {
              "rotate": "0deg",
            },
          ],
        }
      }
    >
      <View
        _text={
          Object {
            "color": "lightText",
          }
        }
        accessible={false}
        bg="rgba(0, 0, 0, 0.3)"
        bottom={0}
        collapsable={false}
        focusable={true}
        importantForAccessibility="no"
        left={0}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onHoverIn={[Function]}
        onHoverOut={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        position="absolute"
        right={0}
        style={
          Array [
            Object {
              "backgroundColor": "rgba(0, 0, 0, 0.3)",
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
          ]
        }
        top={0}
      />
    </View>
    <View
      needsOffscreenAlphaCompositing={true}
      pointerEvents="box-none"
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "transform": Array [
            Object {
              "translateY": 0,
            },
            Object {
              "translateX": 0,
            },
            Object {
              "scale": 1,
            },
            Object {
              "scaleX": 1,
            },
            Object {
              "scaleY": 1,
            },
            Object {
              "rotate": "0deg",
            },
          ],
        }
      }
    >
      <View
        alignItems="center"
        closeOnOverlayClick={true}
        height="100%"
        justifyContent="center"
        pointerEvents="box-none"
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": "100%",
              "justifyContent": "center",
              "width": "100%",
            },
          ]
        }
        width="100%"
      >
        <View
          accessibilityViewIsModal={true}
          aria-modal="true"
          bg="gray.50"
          maxHeight="1184px"
          onAccessibilityEscape={[Function]}
          pl={6}
          pt={6}
          rounded="lg"
          style={
            Array [
              Object {
                "backgroundColor": "#fafafa",
                "borderRadius": 6,
                "maxHeight": 1184,
                "paddingLeft": 24,
                "paddingTop": 24,
                "width": "90%",
              },
              Object {
                "elevation": 4,
                "shadowColor": "#000000",
                "shadowOffset": Object {
                  "height": 2,
                  "width": 0,
                },
                "shadowOpacity": 0.23,
                "shadowRadius": 2.62,
              },
            ]
          }
          width="90%"
        >
          <View
            _text={
              Object {
                "color": "lightText",
              }
            }
            accessibilityLabel="Close dialog"
            accessibilityRole="button"
            accessible={true}
            alignItems="center"
            bg="transparent"
            borderRadius="lg"
            collapsable={false}
            colorScheme="primary"
            flexDirection="row"
            focusable={true}
            icon={
              <Memo
                as={
                  <Icon
                    allowFontScaling={false}
                    name="close"
                    size={12}
                  />
                }
                color="primary.600"
              />
            }
            justifyContent="center"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onHoverIn={[Function]}
            onHoverOut={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            position="absolute"
            right={4}
            size={9}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderRadius": 6,
                  "flexDirection": "row",
                  "height": 36,
                  "justifyContent": "center",
                  "position": "absolute",
                  "right": 16,
                  "top": 16,
                  "width": 36,
                  "zIndex": 1,
                },
              ]
            }
            top={4}
            variant="solid"
            zIndex={1}
          >
            <View
              alignItems="center"
              flexDirection="row"
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                  },
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {},
                  ]
                }
              >
                <RNSVGSvgView
                  accessibilityRole="image"
                  align="xMidYMid"
                  bbHeight={16}
                  bbWidth={16}
                  color={4285624698}
                  focusable={false}
                  meetOrSlice={0}
                  minX={0}
                  minY={0}
                  size={4}
                  stroke=""
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                        "borderWidth": 0,
                      },
                      Array [
                        Object {
                          "color": "#71717a",
                          "height": 16,
                          "width": 16,
                        },
                      ],
                      Object {
                        "flex": 0,
                        "height": 16,
                        "width": 16,
                      },
                    ]
                  }
                  tintColor={4285624698}
                  vbHeight={24}
                  vbWidth={24}
                >
                  <RNSVGGroup
                    propList={
                      Array [
                        "stroke",
                      ]
                    }
                    stroke={null}
                  >
                    <RNSVGGroup>
                      <RNSVGPath
                        d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
                        fill={
                          Array [
                            2,
                          ]
                        }
                        propList={
                          Array [
                            "fill",
                            "stroke",
                          ]
                        }
                        stroke={null}
                      />
                    </RNSVGGroup>
                  </RNSVGGroup>
                </RNSVGSvgView>
              </View>
            </View>
          </View>
          <View
            pb={3}
            pr={6}
            style={
              Array [
                Object {
                  "paddingBottom": 12,
                  "paddingRight": 24,
                },
              ]
            }
          >
            <Text
              color="muted.700"
              fontFamily="body"
              fontSize="xl"
              fontStyle="normal"
              fontWeight="bold"
              style={
                Array [
                  Object {
                    "color": "#404040",
                    "fontSize": 20,
                    "fontStyle": "normal",
                    "fontWeight": "700",
                  },
                ]
              }
            >
              CHANGE_NAME_TITLE
            </Text>
          </View>
          <RCTScrollView>
            <View>
              <View
                pb={7}
                pr={6}
                style={
                  Array [
                    Object {
                      "paddingBottom": 28,
                      "paddingRight": 24,
                    },
                  ]
                }
              >
                <View
                  isInvalid={true}
                  isRequired={true}
                  style={
                    Array [
                      Object {
                        "width": "100%",
                      },
                    ]
                  }
                  width="100%"
                >
                  <View
                    mx={4}
                    style={
                      Array [
                        Object {
                          "marginLeft": 16,
                          "marginRight": 16,
                        },
                      ]
                    }
                  >
                    <TextInput
                      accessibilityDescribedBy="field-react-aria-4783223535-1-feedback"
                      accessibilityInvalid={true}
                      accessibilityRequired={true}
                      accessible={true}
                      allowFontScaling={true}
                      background="transparent"
                      borderColor="error.600"
                      borderRadius="lg"
                      borderWidth={1}
                      color="black"
                      disabled={false}
                      editable={true}
                      fontSize="md"
                      isRequired={true}
                      nativeID="field-react-aria-4783223535-1"
                      onBlur={[Function]}
                      onChangeText={[Function]}
                      onFocus={[Function]}
                      onKeyPress={[Function]}
                      p={2}
                      placeholder="Anzeigename"
                      placeholderTextColor="#a3a3a3"
                      px={4}
                      py={3}
                      readOnly={false}
                      rejectResponderTermination={true}
                      required={true}
                      secureTextEntry={false}
                      style={
                        Array [
                          Object {
                            "borderColor": "#dc2626",
                            "borderRadius": 6,
                            "borderWidth": 1,
                            "color": "#000000",
                            "fontSize": 16,
                            "paddingBottom": 8,
                            "paddingLeft": 8,
                            "paddingRight": 8,
                            "paddingTop": 8,
                          },
                        ]
                      }
                      underlineColorAndroid="transparent"
                      variant="outline"
                    />
                    <View
                      style={
                        Array [
                          Object {},
                        ]
                      }
                    />
                    <View
                      mt={2}
                      nativeID="field-react-aria-4783223535-1-helptext"
                      style={
                        Array [
                          Object {
                            "marginTop": 8,
                          },
                        ]
                      }
                    >
                      <Text
                        color="error.400"
                        fontFamily="body"
                        fontSize="xs"
                        fontStyle="normal"
                        fontWeight={400}
                        style={
                          Array [
                            Object {
                              "color": "#f87171",
                              "fontSize": 12,
                              "fontStyle": "normal",
                              "fontWeight": "400",
                            },
                          ]
                        }
                      >
                        Mindestens 3 Zeichnen und maximal 8 Zeichnen
                      </Text>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </RCTScrollView>
          <View
            flexDirection="row"
            flexWrap="wrap"
            justifyContent="flex-end"
            pr={2}
            py={2}
            style={
              Array [
                Object {
                  "flexDirection": "row",
                  "flexWrap": "wrap",
                  "justifyContent": "flex-end",
                  "paddingBottom": 8,
                  "paddingRight": 8,
                  "paddingTop": 8,
                },
              ]
            }
          >
            <View
              _text={
                Object {
                  "color": "lightText",
                }
              }
              accessibilityRole="button"
              accessible={true}
              alignItems="center"
              bg="primary.500"
              borderRadius="lg"
              collapsable={false}
              colorScheme="primary"
              flexDirection="row"
              focusable={true}
              justifyContent="center"
              ml={3}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onHoverIn={[Function]}
              onHoverOut={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              opacity={0.5}
              px={4}
              py={3}
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#06b6d4",
                    "borderRadius": 6,
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "marginLeft": 12,
                    "opacity": 0.5,
                    "paddingBottom": 12,
                    "paddingLeft": 16,
                    "paddingRight": 16,
                    "paddingTop": 12,
                  },
                ]
              }
              variant="solid"
            >
              <View
                alignItems="center"
                flexDirection="row"
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                    },
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {},
                    ]
                  }
                >
                  <Text
                    color="lightText"
                    fontFamily="body"
                    fontSize="md"
                    fontStyle="normal"
                    fontWeight={600}
                    style={
                      Array [
                        Object {
                          "color": "#FFFFFF",
                          "fontSize": 16,
                          "fontStyle": "normal",
                          "fontWeight": "600",
                        },
                      ]
                    }
                  >
                    SAVE
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
